pkg/identityserver/notification_registry.go

```diff
@@ -256,23 +285,23 @@ func (is *IdentityServer) createNotification(ctx context.Context, req *ttnpb.Cre
 		return nil, err
 	}
 
-	if req.Email && email.GetNotification(ctx, req.GetNotificationType()) == nil {
-		log.FromContext(ctx).WithField("notification_type", req.GetNotificationType()).Warn("email template for notification not registered")
-		req.Email = false
-	}
-
-	receiverUserIDs, err := is.lookupNotificationReceivers(ctx, req)
+	receiverUserIDs, emailReceiverIDs, err := is.lookupNotificationReceivers(ctx, req)
```

Comment:
I'm not sure if I'm a fan of adding an extra return variable to `lookupNotificationReceivers`. From what I can gather we are using the same transaction that we are fetching the `usrReceiverIDs` to fetch the user's notification_preferences.


I would argue that while this is efficient it is a bit confusing in terms of readability. We could open a new transaction to read the users' preferences after emitting the events.

What I mean by this is that we can separate the email and the event emitting logic on the function itself, not even creating a new function/method but just having the logic in separate coding blocks:

```go
func (is *IdentityServer) createNotification(ctx context.Context, req *ttnpb.CreateNotificationRequest) (*ttnpb.CreateNotificationResponse, error) {
	if err := clusterauth.Authorized(ctx); err != nil {
		return nil, err
	}

	receiverUserIDs, err := is.lookupNotificationReceivers(ctx, req)
	if err != nil {
		return nil, err
	}
	notification, err := is.storeNotification(ctx, req, receiverUserIDs...)
	if err != nil {
		return nil, err
	}

	evs := make([]events.Event, 0, len(receiverUserIDs))
	for _, ids := range receiverUserIDs {
		evs = append(evs, evtNotificationCreate.NewWithIdentifiersAndData(ctx, ids, notification))
	}
	events.Publish(evs...)

	// From this point onwards its just the email receivers logic!
	var emailReceiverIDs []*ttnpb.UserIdentifiers
	err = is.store.Transact(ctx, func(ctx context.Context, st store.Store) error {
		// Get the email notification preferences of the receiver users.
		emailReceiverUsers, err := st.FindUsers(ctx, receiverUserIDs, []string{"email_notification_preferences"})
		if err != nil {
			return err
		}
		// Filter only the users that have email notifications enabled for the notification type.
		emailReceiverIDs = filterAllowedEmailReveivers(emailReceiverUsers, req.NotificationType)

		return nil
	})
	if err != nil {
		return nil, err
	}

	if email.GetNotification(ctx, req.GetNotificationType()) == nil {
		log.FromContext(ctx).
			WithField("notification_type", req.GetNotificationType()).
			Warn("email template for notification not registered")
		emailReceiverIDs = nil
	}

	if emailReceiverIDs != nil || len(emailReceiverIDs) > 0 {
		if err := is.SendNotificationEmailToUserIDs(ctx, notification, emailReceiverIDs...); err != nil {
			return nil, err
		}
	}

	return &ttnpb.CreateNotificationResponse{
		Id: notification.Id,
	}, nil
}
```